src/rulesets/Base/indexing.jl

```diff
@@ -81,16 +81,37 @@ For the `rrule` of `y = x[inds...]`, this function is roughly
 `setindex(zero(x), dy, inds...)`, returning the array `dx`.
 Differentiable. Includes `ProjectTo(x)(dx)`.
 """
-function ∇getindex(x::AbstractArray, dy, inds...)
+function ∇getindex(x::AbstractArray{T,N}, dy, inds...) where {T,N}
     # `to_indices` removes any logical indexing, colons, CartesianIndex etc,
     # leaving just Int / AbstractVector of Int
     plain_inds = Base.to_indices(x, inds)
-    dx = _setindex_zero(x, dy, plain_inds...)
-    ∇getindex!(dx, dy, plain_inds...)
-    return ProjectTo(x)(dx)  # since we have x, may as well do this inside, not in rules
+    if plain_inds isa NTuple{N, Int} && T<:Number
+        # scalar indexing
+        return OneElement(dy, plain_inds, axes(x))
+    else  # some from slicing (potentially noncontigous)
+        dx = _setindex_zero(x, dy, plain_inds...)
+        ∇getindex!(dx, dy, plain_inds...)
+        return ProjectTo(x)(dx)  # since we have x, may as well do this inside, not in rules
+    end
 end
 ∇getindex(x::AbstractArray, z::AbstractZero, inds...) = z
 
+"""
+    OneElement(val, ind, axes) <: AbstractArray
+
+Extremely simple `struct` used for the gradient of scalar `getindex`.
+"""
+struct OneElement{T,N,I,A} <: AbstractArray{T,N}
+  val::T
+  ind::I
+  axes::A
+  OneElement(val::T, ind::I, axes::A) where {T<:Number, I<:NTuple{N,Int}, A<:NTuple{N,AbstractUnitRange}} where {N} = new{T,N,I,A}(val, ind, axes)
+end
+Base.size(A::OneElement) = map(length, A.axes)
+Base.axes(A::OneElement) = A.axes
+Base.getindex(A::OneElement{T,N}, i::Vararg{Int,N}) where {T,N} = ifelse(i==A.ind, A.val, zero(T))
+# TODO: should we teach ProjectTo that OneElement is more structurally sparse than anything it intersects nonstructural zeros with?
```

Comment:
Thoughts?

Reply:
Not sure I follow the sentence. 

Is OneElement structurally sparse? The question can only arise for 2nd derivatives, which leads me to thin kit probably is.

Reply:
OneElement is structually sparse yes.
And it is like something that should generally be preserved. Like how every Diagonal is a UpperTrianguar so `ProjectTo(::UpperTriangular)(::Diagonal)::Diagonal`  passes it through.
Same goes for like a OneElement that is on the diagonal